src/V3Randomize.cpp

```diff
@@ -41,6 +41,61 @@
 
 VL_DEFINE_DEBUG_FUNCTIONS;
 
+// ######################################################################
+// Establishes the target of a rand_mode() call
+
+struct RandModeTarget final {
+    // MEMBERS
+    AstVar* receiverp;  // Variable that is the target of the rand_mode() call
+    AstNodeExpr* fromp;  // Expression from which the rand_mode() call originates
+    AstClass* classp;  // Class in which rand_mode should be set
+
+    // METHODS
+    static RandModeTarget get(AstNodeExpr* fromp, AstNodeModule* modp) {
+        if (!fromp) return {nullptr, nullptr, VN_AS(modp, Class)};
+        if (AstCMethodHard* const methodHardp = VN_CAST(fromp, CMethodHard)) {
+            return RandModeTarget::get(methodHardp->fromp(), modp);
+        }
+        AstVar* receiverp = nullptr;
+        if (AstVarRef* const varrefp = VN_CAST(fromp, VarRef)) {
+            receiverp = varrefp->varp();
+        } else if (AstMemberSel* const memberSelp = VN_CAST(fromp, MemberSel)) {
+            receiverp = memberSelp->varp();
+        }
+        UASSERT_OBJ(receiverp, fromp, "Unknown rand_mode() receiver");
+        if (receiverp->isRand()) {
+            if (AstMemberSel* const memberselp = VN_CAST(fromp, MemberSel)) {
+                return {receiverp, memberselp->fromp(),
+                        VN_AS(memberselp->fromp()->dtypep()->skipRefp(), ClassRefDType)->classp()};
+            }
+        } else {
+            AstClassRefDType* const classRefDtp
+                = VN_CAST(receiverp->dtypep()->skipRefp(), ClassRefDType);
+            if (classRefDtp) return {receiverp, fromp, classRefDtp->classp()};
+        }
+        return {receiverp, nullptr, VN_AS(modp, Class)};
+    }
+};
+
+// ######################################################################
+// Stores info about a variable's rand_mode state
+
+struct VarRandMode final {
+    // MEMBERS
+    bool usesRandMode : 1;  // True if variable uses rand_mode
+    uint32_t index : 30;  // Index of var in rand_mode vector
+
+    // METHODS
+    static VarRandMode get(const AstVar* varp) {
+        int user1 = varp->user1();
+        VarRandMode randMode;
+        std::memcpy(&randMode, &user1, sizeof(randMode));
```

Comment:
Feels cleaner if VarRandMode was a union of a uint32_t and the subtypes, then can just do an assignment.